src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -1005,32 +1006,89 @@ and tc_maybe_toplevel_term env (e:term) : term                  (* type-checked
+   *   -- typecheck the scrutinee
+   *   -- typecheck the return annotation
+   *   -- typecheck the branches with
+   *      -- env with expected type unset
+   *      -- a new bv, guard_x, as the scrutinee expression
+   *      -- substituting scrutinee bv by guard_x in the return annotation
+   *      -- in tc_eqn: substituting guard_x with the pattern expression, and ascribing it on the branch exp
+   *   -- if the return annotation was a type:
+   *      -- TcUtil.bind_cases as before
+   *   -- if the return annotation was a computation type:
+   *      -- no need to bind cases, take the comp annotation as is
+   *      -- but add the pattern exhaustiveness check in the guard
+   *      -- need to substitute guard_x by scrutinee bv back
+   *   -- bind the scrutinee computation type with the combined branches (using guard_x as the bv in bind)
+   *)
+
   match (SS.compress top).n with
-  | Tm_match(e1, eqns) ->
-    let env1, topt = Env.clear_expected_typ env in
-    let env1 = instantiate_both env1 in
-    let e1, c1, g1 = tc_term env1 e1 in
-    // GM: for some reason, without this annotation on eqns,
-    //     F* fails to do inference. A uvar at type Type gets
-    //     solved to `list u#0`, which is ill-typed.
-    let e1, c1, (eqns : list<S.branch>) =
-        match TcUtil.coerce_views env e1 c1 with
-        | Some (e1, c1) -> (e1, c1, eqns)
-        | None -> (e1, c1, eqns)
-    in
-    let env_branches, res_t, g1 =
-      match topt with
-      | Some t -> env, t, g1
-      | None ->
-        let k, _ = U.type_u() in
-        let res_t, _, g = TcUtil.new_implicit_var "match result" e1.pos env k in
-        Env.set_expected_typ env res_t, res_t, Env.conj_guard g1 g in
+  | Tm_match(e1, ret_opt, eqns) ->  //ret_opt is the return annotation
+    let e1, c1, g1 = tc_term
+      ({(env |> Env.clear_expected_typ |> fst) with instantiate_imp = true})
+      e1 in
 
-    if Env.debug env Options.Extreme
-    then BU.print1 "Tm_match: expected type of branches is %s\n" (Print.term_to_string res_t);
+    let e1, c1 =
+        match TcUtil.coerce_views env e1 c1 with
+        | Some (e1, c1) -> e1, c1
+        | None -> e1, c1 in
+
+    //ret_or_res_typ has type either<either<typ, comp>, typ>
+    let env_branches, ret_or_res_typ, ret_opt, g1 =
+      match ret_opt with
+      | None ->  //no return annotation, set an expected type either from env or a new uvar
+        (match Env.expected_typ env with
+         | Some t -> env, Inr t, None, g1
+         | None ->
+           let k, _ = U.type_u() in
+           let res_t, _, g = TcUtil.new_implicit_var "match result" e1.pos env k in
+           Env.set_expected_typ env res_t, Inr res_t, None, Env.conj_guard g1 g)
+      | Some (t_or_c, None) ->
+        //typecheck the return annotation and unset expected type in the env if exists
+        //the branch typechecking (tc_eqn) will typecheck the branch with an ascription
+        let env, _ = Env.clear_expected_typ env in
```

Comment:
Shouldn't we check that if an expected type is set in the environment, then the type in the returns annotation is compatible with the environment's expected type?

Otherwise, I wonder what would happen if we had something like

```
(match x returns t with  ...) <: t'
```

Reply:
We do it at the end of `tc_match`, I have also added a comment here to that effect.